### PR TITLE
RWA-3222: Corrected Timezone calculations

### PR DESCRIPTION
**Pull request by adityadwadasi**
**Before creating a pull request make sure that:**

- [ ] commit messages are meaningful and follow good commit message guidelines
- [ ] README and other documentation has been updated / added (if needed)
- [ ] tests have been updated / new tests has been added (if needed)

Please remove this line and everything above and fill the following sections:

### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RWA-3222

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/controllers/ReceivedMessagesHealthController.java [hunk 1]
```diff
@@ -58,12 +61,15 @@ public Health health() {
                 .build();
         }
 
-        LocalDateTime now = LocalDateTime.now(clock).minusHours(1);
-        LocalDateTime nowMinusOneHours = LocalDateTime.now().minusHours(1);
-        log.info("LocalDate Clock time minus hour {}, localDate time minus one hour {}", now, nowMinusOneHours);
+        ZoneId ukTimeZone = ZoneId.of("Europe/London");
+        ZonedDateTime utcDateTime = LocalDateTime.now(clock).atZone(ZoneOffset.UTC);
+        ZonedDateTime zonedDateTime = utcDateTime.withZoneSameInstant(ukTimeZone);
+        LocalDateTime localTime = zonedDateTime.toLocalDateTime();
```
Comment (raghera) [inline src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/controllers/ReceivedMessagesHealthController.java:67]: Probably better to name these as `ukZonedDateTime` and `ukLocalDateTime`?

File: src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/controllers/ReceivedMessagesHealthController.java [hunk 2]
```diff
@@ -58,12 +61,15 @@ public Health health() {
...
 
-        if (isDateWithinWorkingHours(now)) {
-            if (repository.getNumberOfMessagesReceivedInLastHour(now) == 0) {
+        log.info("UTC time {}, localDate time {}", utcDateTime, localTime);
```
Comment (raghera) [inline src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/controllers/ReceivedMessagesHealthController.java:69]: again specify what the it's UK datetime

File: src/test/java/uk/gov/hmcts/reform/wacaseeventhandler/controllers/ReceivedMessagesHealthControllerTest.java [hunk 3]
```diff
@@ -181,10 +197,17 @@ private static Stream<LocalDateTime> nonWorkingHoursScenarioProvider() {
...
+            LocalDateTime.of(2022, Month.SEPTEMBER, 01, 8, 31),
+            LocalDateTime.of(2022, Month.SEPTEMBER, 01, 16, 29),
+            LocalDateTime.of(2022, Month.SEPTEMBER, 01, 16, 30)
+        );
+    }
+
+    private static Stream<LocalDateTime> workingHoursWithTimeZoneScenarioProvider() {
+        return Stream.of(
+            LocalDateTime.of(2024, Month.JANUARY, 01, 10, 00),
+            LocalDateTime.of(2024, Month.MAY, 01, 9, 00)
```
Comment (raghera) [inline src/test/java/uk/gov/hmcts/reform/wacaseeventhandler/controllers/ReceivedMessagesHealthControllerTest.java:210]: I think we may need more tests unless they exist already which is not in this PR?  Let me know if already covered elsewhere.

We want tests where the clock is at UTC/GMT date and the time is within working hours and outside working hours.  We'd want to test the beginning of the day and the end of the day.
Then we also want tests where we are in BST and the time is within working hours and outside working hours.  Again we'd want to test the beginning and end of the day for those.

Also there's a scenario where no messages are received during working hours?

I think here you are covering some of the positive cases but not all.

File: src/integrationTest/java/uk/gov/hmcts/reform/wacaseeventhandler/controllers/ReceivedMessagesHealthControllerTest.java [hunk 4]
```diff
@@ -66,7 +66,7 @@ void setup() {
     @Test
     void test_health_reports_down_if_no_messages_received_in_last_hour_during_working_hours() throws Exception {
         // GIVEN
-        setClock(LocalDateTime.of(2022, 8, 26, 17,15));
+        setClock(LocalDateTime.of(2022, 8, 26, 16,15));
```
Comment (raghera) [inline src/integrationTest/java/uk/gov/hmcts/reform/wacaseeventhandler/controllers/ReceivedMessagesHealthControllerTest.java:69]: Why did you change this one?  Isn't 17.15 still within working hours?

File: src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/controllers/ReceivedMessagesHealthController.java (hunk 2 above)
Comment (adityadwadasi) [reply src/main/java/uk/gov/hmcts/reform/wacaseeventhandler/controllers/ReceivedMessagesHealthController.java]: I changed the variables names to the suggested ones Ravi.

File: src/test/java/uk/gov/hmcts/reform/wacaseeventhandler/controllers/ReceivedMessagesHealthControllerTest.java (hunk 3 above)
Comment (adityadwadasi) [reply src/test/java/uk/gov/hmcts/reform/wacaseeventhandler/controllers/ReceivedMessagesHealthControllerTest.java:210]: The old unit tests in the ReceivedMessagesHealthControllerTest.java covered the scenarios mentioned in your comment Ravi.
Eg: 
1. test_health_does_not_call_repository_if_working_day_time_is_out_of_working_hours
2. test_health_calls_repository_if_working_day_time_is_within_working_hours
3. test_health_reports_error_if_no_messages_received_in_last_hour

**Review by raghera — DISMISSED** [review]

**Review by BenNzewi — DISMISSED** [review]

File: src/test/java/uk/gov/hmcts/reform/wacaseeventhandler/controllers/ReceivedMessagesHealthControllerTest.java (hunk 3 above)
Comment (BenNzewi) [inline src/test/java/uk/gov/hmcts/reform/wacaseeventhandler/controllers/ReceivedMessagesHealthControllerTest.java:210]: What about DayLight Saving time? Will this be handled correctly? 
I.e. March 31st 2024, 17:00pm. Will this test pass?

**Review by raghera — DISMISSED** [review]

File: src/test/java/uk/gov/hmcts/reform/wacaseeventhandler/controllers/ReceivedMessagesHealthControllerTest.java (hunk 3 above)
Comment (adityadwadasi) [reply src/test/java/uk/gov/hmcts/reform/wacaseeventhandler/controllers/ReceivedMessagesHealthControllerTest.java:225]: I added a new nonWorkingHours edge case Ben.

**Review by BenNzewi — APPROVED** [review]
